### PR TITLE
Review gossip block validation

### PR DESCRIPTION
**Pull request by dapplion**
**Motivation**

Apply the same revision and style as https://github.com/ChainSafe/lodestar/pull/2760

**Description**

- Inline spec comments to ensure we cover all validation cases.
- Use a seen cache in memory instead of doing db reads
- Port useful optimizations from Lighthouse gossip validation
- Update test with helpful comments and updated sorted conditions

### CODE REVIEW

File: packages/lodestar/src/chain/seenCache/seenBlockProposers.ts [hunk 1]
```diff
@@ -0,0 +1,34 @@
+import {Epoch, Slot, ValidatorIndex} from "@chainsafe/lodestar-types";
+import {MapDef} from "../../util/map";
+
+/**
+ * Keeps a cache to filter unaggregated attestations from the same validator in the same epoch.
```
Comment (twoeths) [inline packages/lodestar/src/chain/seenCache/seenBlockProposers.ts]: need to fix this description?

Comment (dapplion) [reply packages/lodestar/src/chain/seenCache/seenBlockProposers.ts]: Fixed

File: packages/lodestar/src/chain/validation/block.ts [hunk 2]
```diff
@@ -26,56 +40,85 @@ export async function validateGossipBlock(
...
+  if (parentBlock === null) {
+    // If fork choice does *not* consider the parent to be a descendant of the finalized block,
+    // then there are two more cases:
+    //
+    // 1. We have the parent stored in our database. Because fork-choice has confirmed the
+    //    parent is *not* in our post-finalization DAG, all other blocks must be either
+    //    pre-finalization or conflicting with finalization.
+    // 2. The parent is unknown to us, we probably want to download it since it might actually
+    //    descend from the finalized root.
+    if ((await db.block.get(parentRoot as Uint8Array)) !== null) {
```
Comment (twoeths) [inline packages/lodestar/src/chain/validation/block.ts]: this never happen with the current implementation. When `archiveBlocks` runs, if it found any blocks conflicting to the finalized block it'll delete it from db

Just afraid we query db for nothing here.

Comment (dapplion) [reply packages/lodestar/src/chain/validation/block.ts]: That's a good point, and this check won't save us from blocks from unknown forks if they are two children deep from our latest known forked block. I'll remove

**Review by twoeths — APPROVED** [review]

**Pull request merged**